packages/client/src/runtime.ts

```diff
@@ -70,110 +93,183 @@ export type RuntimeApi = Observable<Runtime> & {
+      getCompatibilityLevelSync(runtime, ctx) >= CompatibilityLevel.Partial
+  }
+  const compatibleRuntime$ = (
+    chainHead: ChainHead$,
+    hash: string | null,
+    error: () => Error,
+  ) =>
+    combineLatest([chainHead.getRuntimeContext$(hash), waitDescriptors()]).pipe(
+      map(([ctx, isCompatible]) => {
+        if (!isCompatible(ctx)) {
+          throw error()
+        }
+        return ctx
+      }),
+    )
+
+  const withCompatibleRuntime =
+    <T>(chainHead: ChainHead$, mapper: (x: T) => string, error: () => Error) =>
+    (source$: Observable<T>): Observable<[T, RuntimeContext]> =>
+      combineLatest([
+        source$.pipe(chainHead.withRuntime(mapper)),
+        waitDescriptors(),
+      ]).pipe(
+        map(([[x, ctx], isCompatible]) => {
           if (!isCompatible(ctx)) {
             throw error()
           }
-          return ctx
+          return [x, ctx]
         }),
       )
 
-    const withCompatibleRuntime =
-      <T>(
-        chainHead: ChainHead$,
-        mapper: (x: T) => string,
-        error: () => Error,
-      ) =>
-      (source$: Observable<T>): Observable<[T, RuntimeContext]> =>
-        combineLatest([
-          source$.pipe(chainHead.withRuntime(mapper)),
-          waitChecksums(),
-        ]).pipe(
-          map(([[x, ctx], isCompatible]) => {
-            if (!isCompatible(ctx)) {
-              throw error()
-            }
-            return [x, ctx]
-          }),
-        )
-
-    return {
-      isCompatible,
-      waitChecksums,
-      withCompatibleRuntime,
-      compatibleRuntime$,
-    }
+  return {
+    getCompatibilityLevel,
+    isCompatible,
+    waitDescriptors,
+    withCompatibleRuntime,
+    compatibleRuntime$,
   }
+}
 export type CompatibilityHelper = ReturnType<typeof compatibilityHelper>
+
+const withOptionalRuntime =
+  <T>(
+    runtimeApi: RuntimeApi,
+    fn: (runtime: Runtime) => T,
+  ): WithOptionalRuntime<T> =>
+  (runtime?: Runtime): any =>
+    runtime ? fn(runtime) : runtimeApi.latest().then(fn)
+export type WithOptionalRuntime<T> = {
+  /**
+   * Returns the result after waiting for the runtime to load.
+   */
+  (): Promise<T>
+  /**
+   * Returns the result synchronously with the loaded runtime.
+   */
+  (runtime: Runtime): T
+}
+
+export interface CompatibilityFunctions {
+  /**
+   * Returns the `CompatibilityLevel` for this call comparing the descriptors
+   * generated on dev time with the current live metadata.
+   */
+  getCompatibilityLevel: WithOptionalRuntime<CompatibilityLevel>
+
+  /**
+   * Checks whether or not the call is compatible with the descriptors generated
+   * on dev time. Returns:
+   * - `true` if they haven't changed or the changes are backwards compatible -
+   * `'partial'` if the call might be compatible depending on the value being
+   * sent or received.
+   * - `false` if the call is incompatible no matter the value being sent or
+   * received.
+   */
+  isCompatible: WithOptionalRuntime<boolean | "partial">
```

Comment:
I would either remove the `isCompatible` API completely, or I would simplify it further so that it just returns a boolean (`true` only if its identical or backwards compatible).

From a DX point of view, having a function that returns a `boolean` or a `partial` is arguably worse than using the `getCompatibilityLevel` directly...

In fact, the more I think about it, the more I think that it should just be removed, as we are always in time to add sugar. However, if we ship this API and then we regret about this decision, then that would be a breaking change.